### ISSUE TITLE
Cross platform remote debugging breakpoint not working (and workaround)

### ISSUE THREAD

**Issue by jackqq** [DonJayamanne/pythonVSCode#252]
I've been trying to debug a hello-world Python program on Linux remotely from Windows using VSCode and this extension. However, breakpoint was not working, caused by 2 problems.

By printing arguments in ptvsd (BreakpointInfo.__init__()) on Linux, I learned that it was instructed to break in 'c:\root\python\test.py' on Linux, which made nonsense.

The local source file path turns out to be translated to remote path at line 367 (convertClientPathToDebugger) of src/client/debugger/Main.ts of this extenstion, but the call to path.resolve assumes we are on Windows, so it adds c: to the beginning.

By changing the call to path.posix.resolve, I came to the workaround to the first problem.
```
return path.posix.resolve(this.attachArgs.remoteRoot, pathRelativeToClientRoot);
```

Then, although the breakpoint was set, ptvsd just wouldn't break there. By printing from ptvsd (DebuggerLoop.loop()), I learned that it was first instructed to 'rest', i.e. resume thread, and then without much chance to catch up, a 'brkt' instruction came, often after the desired breakpoint had passed.

At line 149 (onPythonProcessLoaded) of Main.ts, by changing the behavior in the else clause from sending resume thread to sending stop at entry, I came to the workaround to the second problem.
```
else {
    this.sendEvent(new StoppedEvent("entry", pyThread.Id));
//    this.pythonProcess.SendResumeThread(pyThread.Id);
}
```

Obviously the above two are just workarounds, and would definitely cause other problems, on Windows, of course! I would like to put them forward here as a reference, in hope of helping us to a proper solution. Thanks!

**Comment by DonJayamanne** [comment]
@jackqq , thanks for the great research, i'll look into these today.

**Comment by jackqq** [comment]
Thanks for the fix! The breakpoints now converts to the correct path now.

Sorry to have missed that but however, it is still necessary to sleep for a second at the beginning of the program under debugging, or it would finish even before the breakpoint was set.

Code under debugging:
```python
import ptvsd
ptvsd.enable_attach(secret='my_secret',address = ('0.0.0.0', 3000))
ptvsd.wait_for_attach()

import time
time.sleep(1)
print "Hello, World!"
exit()
```

When I do the following:
* print(inp) in DebuggerLoop.loop()
* print("{0}, {1}".format(filename, lineno)) in BreakpointInfo.__init__()

from ptvsd, as mentioned above, the output is:
```
rest
brkp
/root/python/test-vscode.py, 7
resa
Hello, World!
sehi
```

If the sleep line were commented out, the output would be:
```
rest
brkp
Hello, World! /root/python/test-vscode.py, 7

sehi
```

I'm confused which part sent the 'rest' command since apparently it is instructed to stop on entry.

After all, sleeping for a second is a workaround.

**Comment by DonJayamanne** [comment]
@jackqq , let me get this right:
- Step 1: You have the following code:
```python
import ptvsd
ptvsd.enable_attach(secret='my_secret',address = ('0.0.0.0', 3000))
ptvsd.wait_for_attach()

import time
time.sleep(1)
print "Hello, World!"
exit()
```
- Step 2: You run that on a remote server
- Step 3: You attach the debugger locally
- Step 4: The code runs to completion.

Questions:
- Am I right so far?  
- At what point did you add the break points? Before Step 3 or after Step 3?  
- What is your OS (client and server)

**Comment by jackqq** [comment]
I think my topology can be illustrated as below:
```
+-----------+ (2)
|   ptvsd   +----+ :3000
+-----+-----+    |
      |          |
+-----+-----+    |    +------------------------+
|  centos7  |    |    | vscode with helloworld | (1)
+-----+-----+    |    +---------+--------------+
      |          |              | :4711
+-----+------+   |    +---------+-----------------+
| virtualbox |   +----+ vscode plugin debugServer |
+-----+------+        +----+----------------------+
      |                    |
      +--------------------+
      |       Windows      |
      +--------------------+
```

A virtualbox VM running CentOS 7 with Python 2.7, as the 'server', the Windows host and VS Code as the dev environment. The breakpoint was set at the print(Hello World) line from the top VS Code labeled (1) in the illustration.

I've just also tested on a topology without the debugServer, by directly connecting to port 3000 to the VM, and the result were same:

* When the debuggee (2) starts to run on the VM, it blocks and waits for connection, with nothing output.
* Immediately upon successful attach of the debugger (1), code in the VM continues to run.
* If the sleep workaround is applied, the debuggee (2) blocks after outputting the breakpoint filename and line number, and the debugger (1) breaks at the specified line.
* If the sleep workaround is not applied, the debuggee (2) runs and exits, with all the output.
* The debug console on (1) copies all the output from (2), with an additional error message:

```
There was an error in starting the debug server. Error = {"code":"ECONNRESET","errno":"ECONNRESET","syscall":"read"}
```

**Comment by DonJayamanne** [comment]
I think I got it now. Will look into this tomorrow.

**Comment by DonJayamanne** [comment]
@jackqq 
try the following code:
With the addition of ***ptvsd.break_into_debugger()*** 
This should fix the issue.

```python
import ptvsd
ptvsd.enable_attach(secret='my_secret',address = ('0.0.0.0', 3000))
ptvsd.wait_for_attach()
ptvsd.break_into_debugger()

import time
time.sleep(1)
print "Hello, World!"
exit()
```

**Comment by DonJayamanne** [comment]
The wait_for_attach merely pauses execution until the debugger has attached and then resumes execution. If you want it to break into the debugger immediately after the debugger has attached, or you would like to break into the debugger from anywhere else, then use the ```break_into_debugger``` function.

@jackqq , please do test and let me know how this goes.

**Comment by jackqq** [comment]
Yes, `ptvsd.break_into_debugger()` breaks the program on entry, even without the `time.sleep` workaround. And, by breaking on entry, the desired breakpoint was set, but the first break is now the line after `break_into_debugger` rather than the specified breakpoint.

I guess this solution, more stable and definite though, is analogous to the sleep workaround in effect, by sparing some time for the remote debugger to catch up its breath setting up breakpoints.

The point is after `ptvsd.wait_for_attach()`, the `rest` instruction was issued (not knowing by whom) to continue the program under debug, without giving the remote debugger a chance to set breakpoints in time. Maybe it's a bug (or feature) of ptvsd?

**Comment by DonJayamanne** [comment]
>but the first break is now the line after break_into_debugger rather than the specified breakpoint.

Please ensure you have the same code at both (client and server) ends.

>Maybe it's a bug (or feature) of ptvsd? As in your description:

Yes, unfortuantely ptvsd doesn't seem to support this functionality. Rather odd. I suppose the use of ```break_into_debugger``` is the solution.

@jackqq , thanks for helping me iron out the issues here.

**Comment by jackqq** [comment]
> Please ensure you have the same code at both (client and server) ends.

Yes, I'm pretty sure about this.

> I suppose the use of break_into_debugger is the solution.

Thanks for the standard solution!

**Comment by d4tech** [comment]
I am trying to do the same thing in a Docker container but I am constantly getting:
```
There was an error in starting the debug server. Error = {"code":"ECONNREFUSED","errno":"ECONNREFUSED","syscall":"connect","address":"192.168.99.100","port":9215}
```
I tried running a simple socket script on the same port with the same config and there are no issues while doing that.
Please help.

**Comment by jackqq** [comment]
I'd like to recommend you fire up another issue as this one is closed. Open issues are easier for the developers to track and more likely to get answered/solved. Of course if you think this issue is relevant, it can be referred to in your newly opened issue.

Since you get an ECONNREFUSED error, I guess two possible causes:

1. Firewall may be rejecting incoming connections, but your 'simple socket script' test may have proven me wrong.

2. The debuggee may have resumed running upon debugger attachment, and finished prematurely, while the debugger couldn't catch up to set the breakpoints. So you need a call to `ptvsd.break_into_debugger()` immediately after `wait_for_attach` as mentioned in the above discussion.

**Comment by d4tech** [comment]
@jackqq thanks for the reply, I'll do the same.
